ssl/quic/quic_port.c

```diff
@@ -725,20 +767,127 @@ static int marshal_validation_token(QUIC_VALIDATION_TOKEN *token,
+ *                   set to the required buffer size and function returns
+ *                   immediately.
+ * @param ct_len     Pointer to size_t that will receive the ciphertext length.
+ *                   This also includes bytes for QUIC_RETRY_INTEGRITY_TAG_LEN.
+ *
+ * @return 1 on success, 0 on failure
+ *
+ * The ciphertext format is:
+ * [EVP_GCM_IV_LEN bytes IV][encrypted data][EVP_GCM_TAG_LEN bytes tag]
+ */
+static int encrypt_validation_token(const QUIC_PORT *port,
+                                    const unsigned char *plaintext,
+                                    size_t pt_len,
+                                    unsigned char *ciphertext,
+                                    size_t *ct_len)
+{
+    int iv_len, len, ret = 0;
+    size_t tag_len;
+    unsigned char *iv = ciphertext, *data, *tag;
+
+    if ((tag_len = EVP_CIPHER_CTX_get_tag_length(port->token_ctx)) == 0
+        || (iv_len = EVP_CIPHER_CTX_get_iv_length(port->token_ctx)) <= 0)
+        goto err;
+
+    *ct_len = iv_len + pt_len + tag_len + QUIC_RETRY_INTEGRITY_TAG_LEN;
+    if (ciphertext == NULL) {
+        ret = 1;
+        goto err;
+    }
+
+    data = ciphertext + iv_len;
+    tag = data + pt_len;
+
+    if (!RAND_bytes_ex(port->engine->libctx, ciphertext, iv_len, 0)
+        || !EVP_EncryptInit_ex(port->token_ctx, NULL, NULL, NULL, iv)
+        || !EVP_EncryptUpdate(port->token_ctx, data, &len, plaintext, pt_len)
+        || !EVP_EncryptFinal_ex(port->token_ctx, data + pt_len, &len)
+        || !EVP_CIPHER_CTX_ctrl(port->token_ctx, EVP_CTRL_GCM_GET_TAG, tag_len, tag))
+        goto err;
+
+    ret = 1;
+err:
+    return ret;
+}
+
+/**
+ * @brief Decrypts a validation token using AES-256-GCM
+ *
+ * @param port       The QUIC port containing the decryption key
+ * @param ciphertext The encrypted data (including IV and tag)
+ * @param ct_len     Length of the ciphertext
+ * @param plaintext  Buffer to receive decrypted data. If NULL, pt_len will be
+ *                   set to the required buffer size.
+ * @param pt_len     Pointer to size_t that will receive the plaintext length
+ *
+ * @return 1 on success, 0 on failure
+ *
+ * Expected ciphertext format:
+ * [EVP_GCM_IV_LEN bytes IV][encrypted data][EVP_GCM_TAG_LEN bytes tag]
+ */
+static int decrypt_validation_token(const QUIC_PORT *port,
+                                    const unsigned char *ciphertext,
+                                    size_t ct_len,
+                                    unsigned char *plaintext,
+                                    size_t *pt_len)
+{
+    int iv_len, len = 0, ret = 0;
+    size_t tag_len;
+    const unsigned char *iv = ciphertext, *data, *tag;
+
+    if ((tag_len = EVP_CIPHER_CTX_get_tag_length(port->token_ctx)) == 0
+        || (iv_len = EVP_CIPHER_CTX_get_iv_length(port->token_ctx)) <= 0)
+        goto err;
+
+    /* Prevent decryption of a buffer that is not within reasonable bounds */
+    if (ct_len < iv_len + tag_len || ct_len > ENCRYPTED_TOKEN_MAX_LEN)
```

Comment:
would you mind to add braces here:
`if (ct_len < (iv_len + tag_len) || ct_len > ENCRYPTED_TOKEN_MAX_LEN)`

thanks.